src/include/rho/GCNodeAllocator.hpp

```diff
@@ -0,0 +1,223 @@
+
+#include <cstdint>
+#include <functional>
+
+#include "rho/AddressSanitizer.hpp"
+
+/** Reports an allocation error and calls abort(). */
+void allocerr(const char* message);
+
+namespace rho {
+  // Forward declarations.
+  class AllocationTable;
+  class AllocatorSuperblock;
+  class GCNode;
+
+  /**
+   * Freelists are used to store superblock nodes and large allocations.
+   * For large allocations the block and superblock members are unused.
+   */
+  struct FreeListNode {
+    FreeListNode* m_next;
+
+    /**
+     * The m_superblock pointer is only used for superblock free nodes. It is
+     * used to update the free bitset when this allocation is reclaimed from a
+     * freelist.
+     */
+    AllocatorSuperblock* m_superblock;
+
+    /**
+     * An index into a superblock used to update the free bitset when this
+     * allocation is reclaimed from a freelist.
+     */
+    std::uint32_t m_block;
+  };
+
+  /**
+   * \brief An allocator for allocating and iterating GCNode objects of
+   * different sizes.
+   *
+   * This allocator should only be used to allocate GCNodes, otherwise
+   * you will trick the garbage collector to treat non-GCNodes as GCNodes
+   * during the Sweep phase of garbage collection.
+   */
+  class GCNodeAllocator {
+    public:
+      GCNodeAllocator() = delete;
+
+      /**
+       * \brief Allocate an object of at least the given size.
+       *
+       * This should only be used to allocate memory for rho::GCNode objects.
+       */
+      static void* allocate(std::size_t bytes);
+
+      /** \brief Apply function to all current allocations. */
+      static void applyToAllAllocations(std::function<void(void*)> f);
+
+      /** \brief Free a previously allocated object. */
+      static void free(void* p);
+
+      /** \brief Must be called before any allocations can be made. */
+      static void initialize();
+
+      /**
+       * \brief Find heap allocation start pointer.
+       *
+       * This function finds the corresponding allocation for an internal or
+       * exact object pointer.
+       *
+       * If the pointer is inside the bounds of the small object arena, then
+       * the corresponding superblock is found via pointer manipulation.
+       *
+       * If the pointer is inside the heap bounds, then we iterate over all hash
+       * collisions for the pointer to find the corresponding allocation.
+       */
+      static GCNode* lookupPointer(void* candidate);
+
+      /** \brief Print allocator state summary for debugging. */
+      static void printSummary();
+
+    private:
+      friend class AllocatorSuperblock;
+      friend class AllocationTable;
+      friend struct FreeListNode;
+
+      /** Allocation table for medium and large allocations. */
+      static AllocationTable* s_alloctable;
+
+      /** Number of different object sizes plus one for the small-object superblocks. */
+      static constexpr unsigned s_num_small_pools = (256 / 8) + 1;
```

Comment:
Use parameters, not constants in here and throughout.  e.g.:
    = s_max_small_block_size / s_small_block_spacing + 1;